File: MultiLanguageSandbox/src/test/deps/typescript/16.ts

/**
 * Counts the number of valid coin toss sequences with no consecutive heads in n tosses.
 * 
 * @param n - The number of coin tosses.
 * @returns The count of valid sequences.
 * 
 * Examples:
 * countValidCoinTossSequences(1) returns 2
 * countValidCoinTossSequences(2) returns 3
 */

function countValidCoinTossSequences(n: number): number {
    if (n === 0) return 0;
    if (n === 1) return 2;
    
    let dp0 = 1; // sequences ending with T
    let dp1 = 1; // sequences ending with H
    
    for (let i = 2; i <= n; i++) {
        const newDp0 = dp0 + dp1; // sequences ending with T can follow any previous sequence
        const newDp1 = dp0;       // sequences ending with H must follow a sequence ending with T
        dp0 = newDp0;
        dp1 = newDp1;
    }
    
    return dp0 + dp1;
}
const testCountValidCoinTossSequences = (): void => {
    console.assert(countValidCoinTossSequences(1) === 2);
    console.assert(countValidCoinTossSequences(2) === 3);
    console.assert(countValidCoinTossSequences(3) === 5);
    console.assert(countValidCoinTossSequences(4) === 8); // Additional test
    console.assert(countValidCoinTossSequences(5) === 13); // Additional test
    // Feel free to add more tests here
    console.assert(countValidCoinTossSequences(40) === 267914296); // Additional test
    console.assert(countValidCoinTossSequences(39) === 165580141);
    console.assert(countValidCoinTossSequences(38) === 102334155);
    // Uncomment the next line to log a message on successful test pass
    // console.log("All tests passed!");
};

testCountValidCoinTossSequences();

File: MultiLanguageSandbox/src/test/deps/pascal/16.pas
{$ASSERTIONS ON}
(*
Function that counts the number of valid coin toss sequences
without consecutive heads. It calculates possible combinations
for a given number of tosses.

Parameters:
- n: an integer representing the number of coin tosses.

Returns:
The count of valid sequences as an unsigned 64-bit integer.

Examples:
>>> count_valid_coin_toss_sequences(1)
2
*)

function count_valid_coin_toss_sequences(n: integer): int64;
var
  i: integer;
  dp: array of int64;
begin
  if n = 0 then
    Exit(1);
  if n = 1 then
    Exit(2);
  
  SetLength(dp, n + 1);
  dp[0] := 1;
  dp[1] := 2;
  
  for i := 2 to n do
  begin
    dp[i] := dp[i-1] + dp[i-2];
  end;
  
  count_valid_coin_toss_sequences := dp[n];
end;
procedure test_count_valid_coin_toss_sequences;
begin
  Assert(count_valid_coin_toss_sequences(1) = 2);
  Assert(count_valid_coin_toss_sequences(2) = 3);
  Assert(count_valid_coin_toss_sequences(3) = 5);
  Assert(count_valid_coin_toss_sequences(4) = 8);
  Assert(count_valid_coin_toss_sequences(5) = 13);
  Assert(count_valid_coin_toss_sequences(40) = 267914296);
  Assert(count_valid_coin_toss_sequences(39) = 165580141);
  Assert(count_valid_coin_toss_sequences(38) = 102334155);

  (* Uncomment the line below to add output for successful tests *)
  // WriteLn('All tests passed!');
end;

begin
  test_count_valid_coin_toss_sequences;
end.